File: apps/backend/src/main/java/com/clinica/veterinaria/service/TemplateConsultaService.java
package com.clinica.veterinaria.service;

import com.clinica.veterinaria.dto.TemplateConsultaDTO;
import com.clinica.veterinaria.entity.TemplateConsulta;
import com.clinica.veterinaria.entity.Usuario;
import com.clinica.veterinaria.repository.TemplateConsultaRepository;
import com.clinica.veterinaria.repository.UsuarioRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class TemplateConsultaService {

    private final TemplateConsultaRepository repository;
    private final UsuarioRepository usuarioRepository;

    public List<TemplateConsultaDTO> findAll() {
        return repository.findByActivoTrueOrderByCategoriaAscNombreAsc()
            .stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public List<TemplateConsultaDTO> findByCategoria(String categoria) {
        return repository.findByCategoriaAndActivoTrueOrderByNombreAsc(categoria)
            .stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public List<TemplateConsultaDTO> search(String nombre) {
        return repository.findByNombreContainingIgnoreCaseAndActivoTrue(nombre)
            .stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public TemplateConsultaDTO findById(Long id) {
        TemplateConsulta template = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Template no encontrado"));
        return toDTO(template);
    }

    public TemplateConsultaDTO create(TemplateConsultaDTO dto, Long usuarioId) {
        Usuario usuario = usuarioRepository.findById(usuarioId)
            .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        TemplateConsulta template = TemplateConsulta.builder()
            .nombre(dto.getNombre())
            .descripcion(dto.getDescripcion())
            .categoria(dto.getCategoria())
            .examenFisico(dto.getExamenFisico())
            .diagnostico(dto.getDiagnostico())
            .tratamiento(dto.getTratamiento())
            .observaciones(dto.getObservaciones())
            .activo(true)
            .creadoPor(usuario)
            .vecesUsado(0)
            .build();

        template = repository.save(template);
        log.info("Template de consulta creado: {}", template.getId());
        return toDTO(template);
    }

    public TemplateConsultaDTO update(Long id, TemplateConsultaDTO dto) {
        TemplateConsulta template = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Template no encontrado"));

        template.setNombre(dto.getNombre());
        template.setDescripcion(dto.getDescripcion());
        template.setCategoria(dto.getCategoria());
        template.setExamenFisico(dto.getExamenFisico());
        template.setDiagnostico(dto.getDiagnostico());
        template.setTratamiento(dto.getTratamiento());
        template.setObservaciones(dto.getObservaciones());
        if (dto.getActivo() != null) {
            template.setActivo(dto.getActivo());
        }

        template = repository.save(template);
        log.info("Template de consulta actualizado: {}", id);
        return toDTO(template);
    }

    public void delete(Long id) {
        TemplateConsulta template = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Template no encontrado"));
        template.setActivo(false);
        repository.save(template);
        log.info("Template de consulta desactivado: {}", id);
    }

    public void incrementarUso(Long id) {
        TemplateConsulta template = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Template no encontrado"));
        template.setVecesUsado(template.getVecesUsado() + 1);
        repository.save(template);
    }

    private TemplateConsultaDTO toDTO(TemplateConsulta template) {
        return TemplateConsultaDTO.builder()
            .id(template.getId())
            .nombre(template.getNombre())
            .descripcion(template.getDescripcion())
            .categoria(template.getCategoria())
            .examenFisico(template.getExamenFisico())
            .diagnostico(template.getDiagnostico())
            .tratamiento(template.getTratamiento())
            .observaciones(template.getObservaciones())
            .activo(template.getActivo())
            .usuarioId(template.getCreadoPor() != null ? template.getCreadoPor().getId() : null)
            .usuarioNombre(template.getCreadoPor() != null ? template.getCreadoPor().getNombre() : null)
            .vecesUsado(template.getVecesUsado())
            .createdAt(template.getCreatedAt())
            .updatedAt(template.getUpdatedAt())
            .build();
    }
}

File: apps/frontend/src/features/templates/services/templateConsultaService.ts
import axios from '@core/api/axios';

export interface TemplateConsulta {
  id: string;
  nombre: string;
  descripcion?: string;
  categoria?: string;
  examenFisico?: string;
  diagnostico?: string;
  tratamiento?: string;
  observaciones?: string;
  activo: boolean;
  usuarioId?: string;
  usuarioNombre?: string;
  vecesUsado: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface TemplateConsultaCreate {
  nombre: string;
  descripcion?: string;
  categoria?: string;
  examenFisico?: string;
  diagnostico?: string;
  tratamiento?: string;
  observaciones?: string;
}

export const templateConsultaService = {
  async getAll(): Promise<TemplateConsulta[]> {
    const response = await axios.get<TemplateConsulta[]>('/templates/consultas');
    return response.data.map(t => ({ ...t, id: String(t.id) }));
  },

  async getByCategoria(categoria: string): Promise<TemplateConsulta[]> {
    const response = await axios.get<TemplateConsulta[]>(`/templates/consultas/categoria/${categoria}`);
    return response.data.map(t => ({ ...t, id: String(t.id) }));
  },

  async search(nombre: string): Promise<TemplateConsulta[]> {
    const response = await axios.get<TemplateConsulta[]>('/templates/consultas/search', {
      params: { nombre }
    });
    return response.data.map(t => ({ ...t, id: String(t.id) }));
  },

  async getById(id: string): Promise<TemplateConsulta> {
    const response = await axios.get<TemplateConsulta>(`/templates/consultas/${id}`);
    return { ...response.data, id: String(response.data.id) };
  },

  async create(data: TemplateConsultaCreate): Promise<TemplateConsulta> {
    const response = await axios.post<TemplateConsulta>('/templates/consultas', data);
    return { ...response.data, id: String(response.data.id) };
  },

  async update(id: string, data: TemplateConsultaCreate): Promise<TemplateConsulta> {
    const response = await axios.put<TemplateConsulta>(`/templates/consultas/${id}`, data);
    return { ...response.data, id: String(response.data.id) };
  },

  async delete(id: string): Promise<void> {
    await axios.delete(`/templates/consultas/${id}`);
  },

  async incrementarUso(id: string): Promise<void> {
    await axios.post(`/templates/consultas/${id}/usar`);
  },
};

